src/Adapter/Product/Repository/ProductRepository.php

```diff
@@ -547,17 +563,53 @@ private function loadProduct(Product $product): Product
         return $product;
     }
 
-    public function getProductTaxRulesGroupId(ProductId $productId): TaxRulesGroupId
+    /**
+     * @return ProductId[]
+     */
+    public function findProductIdsWithoutCategories(): array
     {
-        $result = $this->connection->createQueryBuilder()
-            ->addSelect('p.id_tax_rules_group')
-            ->from($this->dbPrefix . 'product', 'p')
-            ->where('id_product = :productId')
-            ->setParameter('productId', $productId->getValue())
-            ->execute()
-            ->fetchOne()
-        ;
+        $results = Db::getInstance()->executeS('
+			SELECT p.`id_product`
+			FROM `' . _DB_PREFIX_ . 'product` p
+			' . Shop::addSqlAssociation('product', 'p') . '
+			WHERE NOT EXISTS (
+			    SELECT 1 FROM `' . _DB_PREFIX_ . 'category_product` cp WHERE cp.`id_product` = p.`id_product`
+			)
+		');
```

Comment:
ok waiting 8.0.x -> develop merge to use the trait.

Reply:
ok so after it was finally merged I checked more in depth and the trait method requires ShopConstraint it means I would need to refacto both categoryDelete handlers to add a shop constraint (which would also introduce BC break), but that also I would basically have to refacto whole categories deletion to support multistore :sweat: That would require additional multistore tests, handle retrieving/deleting correct shops products for categories, deleting categories based on shop constraint etc... Altough it must be done at some point, but Im afraid it will be too much for a single PR :sweat: 
Maybe to avoid polluting the CategoryRepository, I should just leave those plain sql's in the abstract handler as it was?

Reply:
Ok. let's reduc the refacto on this PR then. And maybe you shouldn't even add this in the repository Let's save this refacto for later and keep the handler's modifications to the minimum